endpoint: Only perform full synchronization periodically

joamaki opened the pull request:
Performing a full dump of the policy map on every policy map synchronization
is expensive and is only necessary to catch rare cases where the agent's view
of the policy map has diverged from the kernel's view which should only happen
either due to kernel or other bugs or some other application modifying the
endpoint policy map.

To reduce the cost of synchronizing large endpoint policy maps perform the
full synchronization only periodically. The full synchronization is defaults
to 15 minutes. Configurable with the hidden option
"--policy-map-full-reconciliation-interval".

Fixes: 9dc1350fdb ("endpoint: Enhance policy map sync")

```release-note
endpoint: Only perform the full policy map synchronization periodically (every 15 minutes) to reduce overhead with large endpoint policy maps
```

joamaki committed:
endpoint: Only perform full synchronization periodically

Performing a full dump of the policy map on every policy map synchronization
is expensive and is only necessary to catch rare cases where the agent's view
of the policy map has diverged from the kernel's view which should only happen
either due to kernel or other bugs or some other application modifying the
endpoint policy map.

To reduce the cost of synchronizing large endpoint policy maps perform the
full synchronization only periodically. The full synchronization is defaults
to 15 minutes. Configurable with the hidden option
"--policy-map-full-reconciliation-interval".

Fixes: 9dc1350fdb ("endpoint: Enhance policy map sync")
Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/endpoint/bpf.go:
```diff
@@ -1374,25 +1380,33 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
-				}
-				defer e.unlock()
-				return e.syncPolicyMapWithDump()
+	e.startSyncOnce.Do(func() {
```
Opted to do this via `sync.Once` instead of changing where `syncPolicyMapController` is invoked from as there were multiple places allocating an endpoint (createEndpoint, parseEndpoint, ...) and no definite "start the endpoint" method. So seemed like keeping the original structure was safest.

christarazi replied:
This kind of invocation pattern is what `pkg/trigger` provides, but what it misses is the retrying ability of `pkg/controller`. A bit unfortunate that there's no mechanism to have both, but `sync.Once` works for now.

joamaki replied:
This made me think we should just add this ability to `pkg/controller`. I added a `CreateController` method that creates the controller if it doesn't exist and does nothing if it does. This way we don't need the `sync.Once` etc.
WDYT?

jrajahalme reviewed (approved):
Thanks for fixing this, I had no idea we were issuing a full dump on each endpoint (policy) regeneration!

derailed commented on pkg/endpoint/bpf.go:
```diff
@@ -1312,20 +1312,26 @@ func (e *Endpoint) dumpPolicyMapToMapState() (policy.MapState, error) {
+// syncPolicyMapWithDump is invoked periodically to perform a full reconciliation
+// of the endpoint's PolicyMap against the BPF maps to catch cases where either
+// due to kernel issue or user intervention the agent's view of the PolicyMap
+// state has converged from the kernel. A warning is logged if this method finds
```
nit: diverged?

joamaki replied:
Thanks good catch.

derailed commented on daemon/cmd/daemon_main.go:
```diff
@@ -838,6 +838,10 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.Int(option.PolicyMapEntriesName, policymap.MaxEntries, "Maximum number of entries in endpoint policy map (per endpoint)")
 	option.BindEnv(vp, option.PolicyMapEntriesName)
 
+	flags.Duration(option.PolicyMapFullReconciliationIntervalName, 15*time.Minute, "Interval for full reconciliation of endpoint policy map")
```
nit: should the default be defined as a const?

joamaki replied:
Good question. Half of the flags defined here seem to set the default directly and other half have it in `defaults` package. My personal preference is not to put it to `pkg/defaults` when it's used only here and only have things in `pkg/defaults` that are e.g. shared with operator and thus repeated. 

Don't mind either way. If you think it'd be nicer to have it in `pkg/defaults` I can move it there.

derailed commented on pkg/endpoint/bpf.go:
```diff
@@ -1374,25 +1380,33 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
+		e.controllers.UpdateController(ctrlName,
+			controller.ControllerParams{
+				Group: syncPolicymapControllerGroup,
+				DoFunc: func(ctx context.Context) (reterr error) {
```
Do we need a named return here?

joamaki replied:
Ah yeah, might as well clean that old code up.

derailed reviewed (approved):
@joamaki Nice work! just a couple comments...

christarazi reviewed (commented):
The fix makes sense to me. How do we know that Cilium wasn't relying on this controller running frequently for correct operation?

joamaki committed:
controller: CreateController method for creation without triggering

The UpdateController method performs either a creation, or updating
of a controller. If an update is performed the controller is immediately
triggered. In some cases we want to create the controller once without
triggering or modifying it if it already has been created.

Add a CreateController method for those use cases.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

christarazi commented on pkg/controller/manager.go:
```diff
@@ -81,33 +81,56 @@ func (m *Manager) updateController(name string, params ControllerParams) *manage
+	} else {
+		m.controllers = controllerMap{}
+	}
+	m.createControllerAndUnlock(name, params)
```
Nit: I think I would prefer a `createControllerLocked` and just do the unlocking of the mutex at the caller level, rather than mixing responsibilities.

joamaki replied:
Didn't want to change the locking semantics (e.g. to hold both this manager and the globalStatus manager locks). But looks like we do that in e.g. `RemoveController` anyway. I'll change it to `createControllerLocked`.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -1339,12 +1345,18 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
+func (e *Endpoint) startSyncPolicyMapController() {
 	ctrlName := fmt.Sprintf("sync-policymap-%d", e.ID)
-	e.controllers.UpdateController(ctrlName,
+	e.controllers.CreateController(ctrlName,
```
Super minor nit but it's odd to return a boolean and ignore it. Looking at the line of code here it's not obvious that the result is being ignored. Ideally the underlying function wouldn't return a boolean if it's unused in production code (I see it's used in tests), but this could at least signal that the result is deliberately ignored with a `_ = e.controllers.CreateController(...)`.
